Add docs for primary constructors (#7236)
dart-lang/site-www: src/content/language/primary-constructors.md

:::

## Overview

Primary constructors provide a concise way to declare a class's fields
and its main constructor in a single line.
They reduce the boilerplate of declaring fields, passing parameters,
and assigning them in the constructor body.
This shorthand is purely "syntax sugar"
and does not introduce new runtime semantics.

### Before and after

Consider this traditional class with two fields and a constructor:

<?code-excerpt "language/lib/primary_constructors/point.dart (point)"?>
```dart
// Current syntax.
class Point {
  int x;
  int y;

  Point(this.x, this.y);
}
```

Using a primary constructor makes the same class much more concise:

<?code-excerpt "language/lib/primary_constructors/point.dart (point-primary)" replace="/PointPrimary/Point/g"?>
```dart
// Using a primary constructor.
class Point(var int x, var int y);
```

Declaring a parameter in the primary constructor with `var` or `final`
implicitly induces an instance variable for that parameter.

To ensure this constructor executes on every new instance,
a class, mixin class, or enum with a primary constructor
cannot have any other non-redirecting generative constructors.

> ## Declare parameters

Reviewer: Rename to `Declaring parameters`

Author: Thanks for the suggestion! However, the [Google developer documentation style guide](https://developers.google.com/style/headings#ing-verb-forms) recommends avoiding -ing verb forms as the first word in headings when possible.

How about `Parameter declarations` instead?

Reviewer: The "declaring" in declaring parameters is moreso of a adjective rather than a verb here, but I see that we'd probably still want to avoid that. It's less so about the section being a section for how to declare parameters and more about "declaring parameters" which allows the users to declare fields in parameters.

I don't think `Parameter declarations` is accurate. Hmm, maybe something like "Field declarations in parameters"?

Author: Sounds good! I like "Field declarations in parameters"

> Placing a parameter list just after the class name

Reviewer: Maybe remove line 57-58 and for line 59 start with "Parameters in primary constructors with the `var` or `final` modifier..."

specifies both a constructor and instance variables.
Parameters with the `var` or `final` modifier, called **declaring parameters**,
implicitly induce a field.

Omitting the modifier creates a **non-declaring parameter**,
which does not induce an implicit field.

> Use non-declaring parameters for custom initialization logic.

Reviewer: Not sure if we need this sentence. Non-declaring parameters are just what we're familiar with existing normal constructor parameters. This section might be more confusing and could use some simplification.

Because `final` and `var` modifiers on parameters are reserved exclusively
for declaring parameters in primary constructors,

> you cannot use them on parameters in other kinds of functions.

Reviewer 2: Perhaps move this paragraph down after the code excerpt (the code illustrates things explained above, and does not give any examples of the error mentioned in this paragraph).

in the class header become directly available
for initializing non-late fields in the class body.

This eliminates the need for a separate initializer list
when calculating values for other fields.
It works just as if you were initializing variables
in a traditional constructor's initializer list.

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (initializer-scope)"?>
```dart
class DeltaPoint(final int x, int delta) {
  final int y = x + delta; // Accesses 'x' and 'delta' parameters directly!
}
```

This makes refactoring between traditional and primary constructors
simpler and safer.

## Add constructor bodies

To validate input or perform complex initialization,
you can add a body to the primary constructor inside the class definition.
This body uses the `this` keyword followed by a block:

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (constructor-bodies)" replace="/PointWithBody/Point/g"?>
```dart
class Point(var int x, var int y) {
  this : assert(x >= 0 && y >= 0) {
    print('Point initialized at ($x, $y)');
  }
}
```

This block can specify an initializer list after `this` and/or a function body.

> ## Private named parameters

Reviewer: I don't think this section title is accurate. Private named parameters is specifically for _named_ parameters so we could explain that in here, but I don't think it's the right title for this section with its current contents.

Then with the contents itself, since the parameter isn't named, you might as well have written:
```dart
class User(final String _name);
```
which would declare a field `final String _named` with the value that the user puts in at the call site (e.g. `User("Kallen")`)
You wouldn't need to have a parameter with a different name here.

Reviewer 2: Perhaps this would be more helpful if the section had two examples showing the same outcome achieved in the old/standard way as well as the new 'private named' way:

```dart
// Variant not using a private named parameter.
class User({required String name}) {
  String _name = name;
}

// Variant using a private named parameter.
class User({required var String _name});
```

.. and then mention that the call sites are identical, `User(name: 'John Doe')`.

To use a public parameter to initialize a private field,
use a non-declaring parameter in the primary constructor header
and declare the private field in the class body:

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (private-fields)" replace="/UserWithPrivateField/User/g"?>
```dart
class User(String name) { // 'name' is public
  final String _name = name; // '_name' is private
}
```

Dart also supports private named parameters in traditional constructors.
To learn more, visit
[Private named parameters](/language/constructors#private-named-parameters).

## Empty bodies

> An empty body of a class, mixin class, or extension type (`{}`)

Reviewer: Extensions can also have `;` bodies.

And they don't need to be using a primary constructor. So we can probably remove the `when using a primary constructor` bit.
Although, then perhaps we should reconsider where this `Empty bodies` section should be.

can be replaced by `;` when using a primary constructor.

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (empty-bodies)" replace="/EmptyBodyPoint/Point/g"?>
```dart
class Point(var int x, var int y);
```

## Super parameters

Super parameters can be declared in the same way
as in traditional constructors today:

<?code-excerpt "language/lib/primary_constructors/super_parameters.dart"?>
```dart
class A(final int a);

class B(super.a) extends A;
```

## Constraints and breaking changes

> To avoid unexpected compiler errors and understand the impact of this feature,

Reviewer: I wouldn't say these are unexpected, but maybe instead we can write something like "These are some of the constraints and errors you should be aware of:"

Author: Changing it to : "Keep these constraints and potential errors in mind when using primary constructors:" 

To make the sentence more active.

keep these constraints and edge cases in mind:

> *   **No `late` or `external` fields**: The `late` and `external` modifiers

Reviewer: Perhaps more like "No `late` or `external` parameters in primary constructors"?
Late or external fields are allowed in general.

Reviewer 2: Could also be "A declaring parameter cannot be `late` or `external` (those variables must be declared in the body, like today)."

    are not allowed on parameters in the primary constructor header.
    To use these modifiers, declare the fields in the class body as usual.
*   **Name collisions**: Declaring a parameter in the primary constructor
    with the same name as a method or another field in the class body
    results in a compile-time error.

:::warning
**Important breaking changes and edge cases:**

> *   **Restriction on `final` in normal function parameters**:

Reviewer: Also, `var`.
Enforced with the `var_with_no_type_annotation` lint.
Also this lint and the `avoid_final_parameters` lint only work in 3.12 and below which might be worth mentioning.

    With primary constructors,
    using `final` on formal parameters in normal functions
    becomes a compile-time error.
    It is reserved exclusively for declaring parameters in primary constructors.
    To enforce immutable parameters as a style choice,
    use the [avoid_final_parameters](https://dart.dev/tools/linter-rules/avoid_final_parameters)
    or [parameter_assignments](https://dart.dev/tools/linter-rules/parameter_assignments) linter rules.
*   **The `factory` method edge case**:
    If you have a method named `factory` with no return type
    (for example, `factory() {}`),
    it will be parsed as a factory constructor after primary constructors ship.

>     Ensure your methods have explicit return types to avoid this conflict.

Reviewer 2: Perhaps:
```suggestion
    Make sure such methods have explicit return types to avoid this conflict.
```
Otherwise it might be interpreted to mean that developers must add explicit return types to all methods, which could be a much bigger effort. Also, it would be counterproductive to add explicit return types on instance methods whose return type is obtained by override inference.

## Primary initializer scope

When you use a primary constructor, the parameters you declare
in the class header become directly available
for initializing non-late fields in the class body.

This eliminates the need for a separate initializer list
when calculating values for other fields.
It works just as if you were initializing variables
in a traditional constructor's initializer list.

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (initializer-scope)"?>
```dart
class DeltaPoint(final int x, int delta) {
  final int y = x + delta; // Accesses 'x' and 'delta' parameters directly!
}
```

This makes refactoring between traditional and primary constructors
simpler and safer.

## Add constructor bodies

To validate input or perform complex initialization,
you can add a body to the primary constructor inside the class definition.
This body uses the `this` keyword followed by a block:

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (constructor-bodies)" replace="/PointWithBody/Point/g"?>
```dart
class Point(var int x, var int y) {
  this : assert(x >= 0 && y >= 0) {
    print('Point initialized at ($x, $y)');
  }
}
```

> This block can specify an initializer list after `this` and/or a function body.

Reviewer 2: Perhaps extend the text slightly to cover forms like `this : assert(x >= 0);` and `@metadata this;` as well?

---
title: Primary constructors

> description: Create classes concisely by declaring parameters and fields in the class header.

Reviewer 3: To keep this under 80 characters:

```suggestion
description: >-
    Create classes concisely by
    declaring parameters and fields in the class header.
```

prevpage:
  url: /language/constructors
  title: Constructors
nextpage:
  url: /language/methods
  title: Methods
---

> :::version-note
> Primary constructors require a language version of at least 3.13.
> :::

Reviewer 3: ```suggestion
:::version-note
Primary constructors require a [language version][] of at least 3.13.
:::

[language version]: /language/versioning
```

## Overview

Primary constructors provide a concise way to declare a class's fields
and its main constructor in a single line.
They reduce the boilerplate of declaring fields, passing parameters,
and assigning them in the constructor body.

> This shorthand is purely "syntax sugar"
> and does not introduce new runtime semantics.

Reviewer 3: Maybe rather than put "syntax sugar" in quotes as a sort of explanation, we could just be more direct about what that means?

```suggestion
This shorthand changes how you write the declaration,
but it doesn't change runtime behavior.
```

## Primary initializer scope

When you use a primary constructor, the parameters you declare
in the class header become directly available
for initializing non-late fields in the class body.

This eliminates the need for a separate initializer list
when calculating values for other fields.
It works just as if you were initializing variables
in a traditional constructor's initializer list.

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (initializer-scope)"?>
```dart
class DeltaPoint(final int x, int delta) {
  // Accesses 'x' and 'delta' parameters directly!
  final int y = x + delta;
}
```

This makes refactoring between traditional and primary constructors
simpler and safer.

## Add constructor bodies

To validate input or perform complex initialization,
you can add a body to the primary constructor inside the class definition.
This body uses the `this` keyword followed by a block:

<?code-excerpt "language/lib/primary_constructors/primary_constructors.dart (constructor-bodies)" replace="/PointWithBody/Point/g"?>
```dart
class Point(var int x, var int y) {
  this : assert(x >= 0 && y >= 0) {
    print('Point initialized at ($x, $y)');
  }
}
```

> This block can specify an initializer list after `this` and/or a function body.

Reviewer 3: It's this type of block, not necessarily this exact block right? Perhaps "the" can expand it to similar situations.

```suggestion
The block can specify an initializer list after `this` and/or a function body.
```

It also supports forms with just an initializer list followed by a semicolon
(for example, `this : assert(x >= 0);`),
and you can apply metadata annotations to it

> (for example, `@metadata this;`).

Reviewer 3: Instead of talking about what it supports, would it make sense to focus on how to complete certain actions/tasks?

```suggestion
To provide only an initializer list,
end it with a semicolon, such as `this : assert(x >= 0);`.
You can also annotate the block with metadata,
such as `@metadata this;`.
```

## Initialize private fields

To initialize a private field using a named parameter,
you can write manual assignment boilerplate in a traditional constructor:

<?code-excerpt "language/lib/primary_constructors/private_named_parameters.dart (private-named-old)" replace="/UserOld/User/g"?>
```dart
// Variant not using a private named parameter.
class User({required String name}) {
  String _name = name;
}
```

With primary constructors and the private named parameters feature,
you can declare the private field directly in the constructor header.
When you use a private name (with a leading underscore) for a named parameter,
the compiler automatically makes the parameter name public for the caller
by removing the underscore:

<?code-excerpt "language/lib/primary_constructors/private_named_parameters.dart (private-named-new)" replace="/UserNew/User/g"?>
```dart
// Variant using a private named parameter.
class User({required var String _name});
```

In both cases, the caller uses the public name `name`
at the call site: `User(name: 'John Doe')`.

## Empty bodies

An empty body of a class, mixin class, extension, or extension type (`{}`)
can be replaced by a semicolon (`;`).

> While this is true in general for these declarations, it is particularly
> useful when using a primary constructor to keep the entire declaration on a single line.

Reviewer 3: ```suggestion
While this is true in general for these declarations,
it is particularly useful when using a primary constructor to
keep the entire declaration on a single line.
```

<?code-excerpt "language/lib/primary_constructors/super_parameters.dart"?>
```dart
class A(final int a);

class B(super.a) extends A;
```

## Constraints and breaking changes

Keep these constraints and potential errors in mind
when using primary constructors:

*   **Declaring parameters cannot be `late` or `external`**:
    The `late` and `external` modifiers are not allowed
    on parameters in the primary constructor header.
    To use these modifiers, declare the fields in the class body as usual.
*   **Name collisions**: Declaring a parameter in the primary constructor
    with the same name as a method or another field in the class body
    results in a compile-time error.

:::warning
**Important breaking changes and edge cases:**

*   **Restriction on `final` and `var` in normal function parameters**:
    With primary constructors,
    using `final` or `var` on formal parameters in normal functions
    becomes a compile-time error.
    They are reserved exclusively for declaring parameters
    in primary constructors.
    Note that the lints `avoid_final_parameters` and `var_with_no_type_annotation`
    only work in Dart 3.12 and below.

>     To enforce immutable parameters as a style choice in Dart 3.13 and later,
>     use the
>     [parameter_assignments](https://dart.dev/tools/linter-rules/parameter_assignments)
>     linter rule.
> *   **The `factory` method edge case**:
>     If you have a method named `factory` with no return type
>     (for example, `factory() {}`),
>     the compiler parses it as a factory constructor
>     after primary constructors ship.
>     Make sure such methods have explicit return types
>     to avoid this conflict.
> :::

Reviewer 3: Let's put the lint in code font and use a link definition.

```suggestion
    To enforce immutable parameters as a style choice in Dart 3.13 and later,
    use the [`parameter_assignments`][] linter rule.
*   **The `factory` method edge case**:
    If you have a method named `factory` with no return type
    (for example, `factory() {}`),
    the compiler parses it as a factory constructor
    after primary constructors ship.
    Make sure such methods have explicit return types
    to avoid this conflict.
:::

[`parameter_assignments`]: /tools/linter-rules/parameter_assignments
```
